src/extensions/data_array.js

```diff
@@ -49,11 +58,31 @@ export default () => {
         });
       };
 
+      // Validate input.
       // 2) Validate that each object in the input has the required properties.
       validateRequiredFields(body, [datastream, dataArray]);
       validateRequiredFields(body.components, [phenomenonTime, result]);
 
-      // 3) Validate that `body.components` and `body.dataArray` are the same
+      // Validate input.
+      // 3) Validate that the 'components' are one of:
+      //    'result', 'phenomenonTime', 'FeatureOfInterest/id'
+      body.components.forEach(component => {
+        switch(component) {
+          case result:
+          case phenomenonTime:
+          case featureOfInterestId:
+            break;
+          default:
+          throw Object.create({
+            name: ERR.BAD_REQUEST,
+            errno: ERR.ERRNO_BAD_REQUEST,
+            errors: 'Invalid component: ' + component
+          });
+        }
+      });
+
+      // Validate input.
+      // 4) Validate that `body.components` and `body.dataArray` are the same
```

Comment:
We also need to validate that `body.dataArray.length == body.dataArrayIotCount` if `dataArray@iot.count` is present in the body.